src/java/htsjdk/samtools/BAMIndexer.java

```diff
@@ -71,14 +67,28 @@ public BAMIndexer(final File output, final SAMFileHeader fileHeader) {
      * @param fileHeader header for the corresponding bam file.
      */
     public BAMIndexer(final OutputStream output, final SAMFileHeader fileHeader) {
+        this(fileHeader, numRefs -> new BinaryBAMIndexWriter(numRefs, output));
+    }
 
+    /*
+     * Prepare to index a BAM.
+     *
+     * @param fileHeader header for the corresponding bam file.
+     * @param  createWrite a lambda that, given an Integer numReferences value, will create a BinaryBAMIndexWriter
+     *                     with that value and an appropriate output.
+      */
+    private BAMIndexer(final SAMFileHeader fileHeader, Function<Integer, BinaryBAMIndexWriter> createWriter) {
```

Comment:
I did try that; you can't create the BAMIndexWriter until you have the number of references, but you can't get that until the common code runs. So using the lambda allows deferral of the BAMIndexWriter creation with the appropriate output type until the end of the execution of  the common code, when we have the refs. I think its either that or duplicate code. Let me know if you see another way.